Add stack-trace docs (#1336)
wpilibsuite/frc-docs: source/docs/software/troubleshooting/robots-shouldnt-quit.rst

Robots Should Not Quit, But Yours Did!
======================================

``Robots should not quit, but yours did!``

> When your robot code hits an unexpected error, you will see this message show up in some console output (Driver Station or rioLog). You'll probably also notice your robot abruptly stop, or possibly never move. These unexpected errors are called *unhandled exceptions*.

Reviewer: I don't think this capitalization of RioLog is correct.
![image](https://user-images.githubusercontent.com/13039555/116152343-ccc8ca80-a6b3-11eb-8013-4880d6909d05.png)
```suggestion
When your robot code hits an unexpected error, you will see this message show up in some console output (Driver Station or RioLog). You'll probably also notice your robot abruptly stop, or possibly never move. These unexpected errors are called *unhandled exceptions*.

```

When an unhandled exception occurs, it means that your code has one or more bugs which need to be fixed.

This article will explore some of the tools and techniques involved in finding and fixing those bugs.

What's a Stack Trace?
---------------------

The ``Robots should not quit`` message is a signal that a *stack trace* has been printed out. 

In C++ and Java, a `stack <https://en.wikipedia.org/wiki/Call_stack>`_ data structure is used to store information about which function or method is currently being executed.

> A *stack trace* prints information about what was on this stack when the unhandled exception occurred.This points you to the lines of code which were running just before the problem happened. While it doesn't always point you to the *exact root cause* of your issue, it's usually the best place to start looking.

Reviewer: ```suggestion
A *stack trace* prints information about what was on this stack when the unhandled exception occurred. This points you to the lines of code which were running just before the problem happened. While it doesn't always point you to the *exact root cause* of your issue, it's usually the best place to start looking.

```

> A *stack trace* prints information about what was on this stack when the unhandled exception occurred.This points you to the lines of code which were running just before the problem happened. While it doesn't always point you to the *exact root cause* of your issue, it's usually the best place to start looking.

Reviewer: Is the emphasis on `exact root cause` necessary/consistent with the rest of the doc writing style?

Author: I'll got take a peek at the style guide, I'm not entirely sure. I'm definitely not tied to having it there, it's more a reflection of where I'd place my verbal emphasis if I were saying it out loud. 

My end goal would really just be to drive home the main point of the whole article: a stack trace is a _tool to solve your problem_, not a solution. I've found new folks often get confused (and discouraged) when they see an experienced person look at a stack trace (or similar), come up with an answer, and assume that the answer must have come from within said stack trace.

TL;DR probably doesn't have to be there.

Reviewer: Okay, fair enough :)

> What's an "Unhandled Exception"?

Reviewer: Quotes here aren't consistent with the other header.

Search for More Information
^^^^^^^^^^^^^^^^^^^^^^^^^^^

`Google <https://google.com>`_ is a phenomenal resource for understanding the root cause of errors. Searches involving the programming language and the name of the exception will often yield good results on more explanations for what the error means, and how it often comes about.

Seeking Outside Help
^^^^^^^^^^^^^^^^^^^^

If all else fails, you can seek out advice and help from others (both in-person and online). When working with folks who aren't familiar with your codebase, it's very important to provide the following information:

 * Access to your source code, (EX: :ref:`on github.com <docs/software/basic-programming/git-getting-started:Git Version Control Introduction>`)
 * The **full text** of the error, including the full stack trace.

Common Examples & Patterns
--------------------------

There are a number of common issues which result in runtime exceptions. 

> ``Null``/``NULL``

Reviewer: If possible, this might be better as:
```suggestion
``null``/``NULL``
```

Author: Lemme go look, this might get resolved as part of the C++ expansion. You got it though, the intent was to cover Java's `null` and C++'s `NULL`. I'm thinking an `inline code block containing NULL` will indicate the programming-language specific identifier, while plain old "null" will indicate the general concept.

Reviewer: Ah, okay. In any case, if you want to talk about the concept of null values, that wouldn't be in a code tag, e.g.
> Null/`null`/`NULL`
although I personally would prefer just `null`/`NULL`.

^^^^^^^^^^^^^^^^^

> Both C++ and Java have the concept of "null" - a reference which has not yet been initialized, and does not refer to anything meaningful.

Reviewer: I'm not too keen on this explanation, since it might be the programmer's first introduction to `null`, which this doesn't give a broad enough description of IMO. Perhaps:
```suggestion
Both C++ and Java have the concept of "null", a value that does not refer to anything meaningful. In Java, `null` is assigned when there is no value given to a variable declared in a method, or a non-primitive (e.g. not an instance of a class) member of a class. In C++, TODO
```

By going to line 24, you can see there is only one thing which could be null - ``armMotorCtrl``. Looking further up, you can see that the ``armMotorCtrl`` object is declared, but never instantiated.

Alternatively, you can step through lines of code with the single step debugger, and stop when you hit line 24. Inspecting the ``armMotorCtrl`` object at that point would show that it is null. 

Fixing Null Object Issues
"""""""""""""""""""""""""

Generally, you will want to ensure each reference has been initialized before using it. In this case, there is a missing line of code to instantiate the ``armMotorCtrl`` before calling the ``setInverted()`` method. 

A functional implementation could look like this: 

.. tabs::

   .. group-tab:: Java

      .. code-block:: Java
          :lineno-start: 19

            PWMSparkMax armMotorCtrl;

            @Override
            public void robotInit() {
                
                armMotorCtrl = new PWMSparkMax(0);
                armMotorCtrl.setInverted(true);

            }

   .. group-tab:: C++

      .. code-block:: C++

         //TODO

Divide by Zero
^^^^^^^^^^^^^^

> It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the RoboRIO) will cause an Unhandled Exception.

Reviewer: ```suggestion
It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the roboRIO) will cause an Unhandled Exception.
```

> It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the RoboRIO) will cause an Unhandled Exception.

Reviewer: As far as wording goes, I say the processor itself isn't causing the exception.
```suggestion
It is not generally possible to divide an integer by zero, and expect reasonable results. Most processors (including the RoboRIO) will raise an Unhandled Exception.
```

Author: Agreed, "Raise" is better here. 

I'll do some thinking as well as we start to reword for C++ too, as the core concept is straighforward, but nuanced:

The _problem_ is some logical inconsistency in the user's code, which puts the processor (or VM) in a state where it doesn't make sense to continue executing.
The exception users observe is a _result_ of the processor, the VM, or some code going into "we have a problem" mode, stopping the normal flow of execution, and raising the exception up.

               at edu.wpi.first.wpilibj.RobotBase.lambda$startRobot$0(RobotBase.java:387)
               at java.base/java.lang.Thread.run(Thread.java:834)

         Warning at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:350): Robots should not quit, but yours did!
         Error at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:352): The startCompetition() method (or methods called by it) should have handled the exception above.

   .. group-tab:: C++

      .. code-block:: text

          TODO

Looking at the stack trace, we can see a ``java.lang.ArithmeticException: / by zero`` exception has occurred on line 25. If you look at the two variables which are used on the right-hand side of the ``=`` operator, you might notice one of them has not been initialized. This means its value is, by default, zero. And, the zero-value variable is used in the denominator of a division operation. Hence, the divide by zero error happens.

Alternatively, by running the single-step debugger and stopping on line 25, you could inspect the value of all variables to discover ``shoulderToElbow_in`` has a value of ``0``.

> Fixing Div/0 Issues

Reviewer: Unnecessary abbreviation?

Author: Possibly. Lemme go check the exact text of what gets printed out in C++/Java statements. The only goal of using an alternate way to "spell" the same concept is to enhance searchability.

Reviewer: Ah okay. If it is using the same phrasing as the error message, then maybe consider putting the "Div/0" in a code block or italics.

               at edu.wpi.first.wpilibj.PWM.<init>(PWM.java:51)
               at edu.wpi.first.wpilibj.PWMSpeedController.<init>(PWMSpeedController.java:20)
               at edu.wpi.first.wpilibj.PWMSparkMax.<init>(PWMSparkMax.java:31)
               at frc.robot.Robot.robotInit(Robot.java:26)
               at edu.wpi.first.wpilibj.TimedRobot.startCompetition(TimedRobot.java:94)
               at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:335)
               at edu.wpi.first.wpilibj.RobotBase.startRobot(RobotBase.java:407)
               at frc.robot.Main.main(Main.java:23)

         Warning at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:350): Robots should not quit, but yours did!
         Error at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:352): The startCompetition() method (or methods called by it) should have handled the exception above.

   .. group-tab:: C++

      .. code-block:: text

          TODO

This stack trace shows that a``edu.wpi.first.hal.util.UncleanStatusException`` has occurred. It also gives the helpful message: ``HAL: Resource already allocated``.

> Looking at our stack trace, we see that the error *actually* happened deep within some WPILib content. However, we should start by looking in our own code. Halfway through the stack trace, you can find a reference to the last line of the team's robot code that called into WPILib: ``Robot.java:26``.

Reviewer: For brevity.
```suggestion
Looking at our stack trace, we see that the error *actually* happened deep within WPILib. However, we should start by looking in our own code. Halfway through the stack trace, you can find a reference to the last line of the team's robot code that called into WPILib: ``Robot.java:26``.
```

> Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefor, the WPILib libraries purposefully generate a custom error message and exception to alert the software developers of a non-achievable hardware configuration.

Reviewer: ```suggestion
Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefore, the WPILib libraries purposefully generate a custom error message and exception to alert the software developers of a non-achievable hardware configuration.
```

> Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefor, the WPILib libraries purposefully generate a custom error message and exception to alert the software developers of a non-achievable hardware configuration.

Reviewer: ```suggestion
Taking a peek at the code, we see line 26 is where the second motor controller is declared. We can also note that *both* motor controllers are assigned to PWM output ``0``. This doesn't make logical sense, and isn't physically possible. Therefor, WPILib purposefully generates a custom error message and exception to alert the software developers of a non-achievable hardware configuration.
```
